packages/malloy-db-duckdb/src/duckdb_connection.ts

```diff
@@ -43,22 +43,28 @@ const duckDBToMalloyTypes: { [key: string]: AtomicFieldTypeInner } = {
 };
 
 export class DuckDBConnection implements Connection, PersistSQLResults {
-  public readonly name;
-
   protected connection;
   protected database;
   protected isSetup = false;
 
   constructor(
-    name: string,
+    public readonly name: string,
     databasePath = "test/data/duckdb/duckdb_test.db",
     workingDirectory = "/"
   ) {
-    this.name = name;
-
-    // TODO temp! For now, just connect to the test database
-    this.database = new Database(databasePath, OPEN_READONLY);
+    this.database = new Database(
+      databasePath,
+      databasePath === ":memory:" ? OPEN_READWRITE : OPEN_READONLY,
+      (err) => {
+        if (err) {
+          return console.error(err);
```

Comment:
Unfortunately since this is an async callback we can't really do anything about errors that happen here, we should divide creating connections into construction and initialization so we can handle any errors that happen during setup.

Reply:
Makes sense. Do you think we should do that now?

Reply:
Maybe something like this (in the `connections.ts` file in the composer demo):

```ts
class ConnectionMap implements LookupConnection<TestableConnection> {
  private readonly connections: Map<string, TestableConnection> = new Map();

  private async createConnection(connectionName: string) {
    let connection: TestableConnection;
    switch (connectionName) {
      case "bigquery":
        connection = new BigQueryConnection("bigquery");
        break;
      case "postgres":
        connection = new PostgresConnection("postgres");
        break;
      case "duckdb":
        connection = new DuckDBConnection("duckdb", ":memory:");
        break;
      default:
        throw new Error(`No connection named ${connectionName}`);
    }
    await connection.initialize();
    return connection;
  }

  async lookupConnection(
    connectionName = "bigquery"
  ): Promise<TestableConnection> {
    let connection = this.connections.get(connectionName);
    if (connection === undefined) {
      connection = await this.createConnection(connectionName);
      this.connections.set(connectionName, connection);
    }
    return connection;
  }
}

export const CONNECTIONS = new ConnectionMap();
```